torch/_refs/nn/functional/__init__.py

```diff
@@ -294,3 +295,37 @@ def tanhshrink(a: TensorLikeType) -> TensorLikeType:
             "Expected a tensor input for an elementwise unary operation!"
         )
     return refs.sub(a, refs.tanh(a))
+
+
+@elementwise_unary_scalar_wrapper
+@elementwise_type_promotion_wrapper(
+    type_promoting_args=("a"),
+    type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT,
+)
+def hardtanh(
+    a: TensorLikeType,
+    min_val: NumberType = -1.0,
+    max_val: NumberType = 1.0,
+    inplace: bool = False,
+) -> TensorLikeType:
+    """
+    Reference implementation of torch.nn.functional.hardtanh
+    """
+    if inplace:
+        raise NotImplementedError
+    if not isinstance(a, TensorLike):
+        raise RuntimeError(
+            "Expected a tensor input for an elementwise unary operation!"
+        )
+    if utils.is_boolean_dtype(a.dtype):
+        raise RuntimeError("Bool inputs not supported for hardtanh")
+
+    # preserve legacy behavior of boundaries not causing type promotion
+    if utils.is_integer_dtype(a.dtype):
+        min_val = int(min_val)
+        max_val = int(max_val)
+        if not (a.dtype != torch.uint8 or (min_val >= 0 and max_val >= 0)):
+            raise RuntimeError(
+                "Cannot do hardtanh on an unsigned type with negative limits"
+            )
+    return torch.clamp(a, min_val, max_val)
```

Comment:
This is so weird, so not only does hardtanh say it has this behavior for legacy reasons:

https://github.com/pytorch/pytorch/blob/369d9f4137a8bfc20e6a4e1d6ab35eeae4e9b345/aten/src/ATen/native/Activation.cpp#L373

but it has its own custom backward:

https://github.com/pytorch/pytorch/blob/8da4993557029d74d88a2641ae5c74d1abfc96bc/aten/src/ATen/native/cuda/Activation.cu#L351

even though the entire thing should just be a call to clamp. Looks like @ngimel worked on this most recently.

So here are my concerns:

- will differentiating this work correctly without registering a custom backward for this reference?
- should we file an issue that updates hardtanh's behavior to be consistent with other elementwise operators? That would mean that integer tensors with float min_val or max_val would throw an error, which seems preferable to silently converting the values to int

Reply:
1) This is a fair question for all ops we are decomposing, and the answer is "who knows". But if AOT tracing is done before decomposition, it doesn't matter, we just need to write decompositions for backwards functions also (which is POR)
2) It won't work without a lot of changes, default values are float, so we'd need to change defaults too. Given that legacy behavior didn't cause any problems or confusion, I propose leaving it as is.

Reply:
I mean, that backwards formula looks *almost* identical to the backwards formula for clamp? https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/FunctionsManual.cpp#L768

Only difference seems that `clamp_backward` takes the inverse conditional, but I don't think that should matter for nan propagation.

> should we file an issue that updates hardtanh's behavior to be consistent with other elementwise operators?

I'm fine with this :)

Reply:
(1) is PoR for PyTorch 2 tracing but the primTorch program (which was before and will be after PyTorch 2) wants to be able to differentiate things eagerly, too

I'm OK with leaving this as-is as long as we file and cite an issue, and if it affects gradient computation then it might be easier to fix that issue than write a custom grad formula

Reply:
Filed an issue here: https://github.com/pytorch/pytorch/issues/78786